nixos/modules/system/activation/switch-to-configuration.py

```diff
@@ -0,0 +1,611 @@
+            transitions.start_unit(unit)
+        elif prev.options != new.options:
+            # Mount options changes, so remount it.
+            transitions.reload_unit(unit)
+
+    # Also handles swap devices.
+    for device in prev_swaps.keys():
+        if device not in new_swaps:
+            # Swap entry disappeared, so turn it off.  Can't use
+            # "systemctl stop" here because systemd has lots of alias
+            # units that prevent a stop from actually calling
+            # "swapoff".
+            print_err("stopping swap device:", device)
+            subprocess.check_call(["@utillinux@/sbin/swapoff", device])
+
+        # FIXME: update swap options (i.e. its priority).
+
+def reload_user_units() -> None:
+    args = ["@systemd@/bin/loginctl", "list-users", "--no-legend"]
+    active_users = subprocess.check_output(args).split('\n')
+    for f in active_users:
+        m = re.match(r"^\s*(?P<qid>\d+)\s+(?P<user>\S+)", f)
+        if m is None:
+            continue
+        uid = m.group(1)
+        user = m.group(2)
+        print_err("reloading user units for {user}...".format(user=user))
+
+        subprocess.call(["@su@", "-s", "@shell@", "-l", user,
+                        "-c", "XDG_RUNTIME_DIR=/run/user/{uid} @systemd@/bin/systemctl --user daemon-reload".format(uid=uid)])
+        subprocess.call(["@su@", "-s", "@shell@", "-l", user,
+                        "-c", "XDG_RUNTIME_DIR=/run/user/{uid} @systemd@/bin/systemctl --user start nixos-activation.service".format(uid=uid)])
+
+
+def main() -> None:
+    if len(sys.argv) < 1:
+        die_with_usage()
+
+    action = sys.argv[1]
+    valid_modes = ["switch", "boot", "test", "dry-activate"]
+    if action not in valid_modes:
+        die_with_usage()
+
+    if not is_nixos():
+        die("This is not a NixOS installation!")
+
+    openlog("nixos")
+
+    # Install or update the bootloader.
+    if action == "switch" or action == "boot":
+        subprocess.check_call(["@installBootLoader@", out])
+
+    # Just in case the new configuration hangs the system, do a sync now.
+    if os.environ.get('NIXOS_NO_SYNC') != "1":
+        subprocess.check_call(["@coreutils@/bin/sync", "-f", "/nix/store"])
+
+    if action == "boot":
+        sys.exit(0)
+
+    # Check if we can activate the new configuration.
+    old_version = read_if_exists(Path("/run/current-system/init-interface-version"), otherwise="")
+    new_version = (out / "init-interface-version").read_text()
+
+    if new_version != old_version:
+        print_err("""
+            Warning: the new NixOS configuration has an ‘init’ that is
+            incompatible with the current configuration.  The new configuration
+            won\'t take effect until you reboot the system.
+        """)
+        sys.exit(100)
+
+    # Ignore SIGHUP so that we're not killed if we're running on (say)
+    # virtual console 1 and we restart the "tty1" unit.
+    signal.signal(signal.SIGPIPE, signal.SIG_IGN)
```

Comment:
SIGHUP (according to the comment) or SIGPIPE (according to the code) ?

Reply:
That is a good question; unfortunately this inconsistency was also present in the Perl implementation. I suspect we would want SIGHUP here.